### PR TITLE
sql: enable by-ID and by-name references to tenants

### LINKED ISSUE #89967: streamingest: make the name of the target tenant customizable
as part of the UA work we are giving names to tenants.  (#89966)

Currently in `streamingccl > streamingest >  stream_ingestion_planning.go` the TenantInfo of the new target tenant is created with no name.

It is important that this name becomes customizable, for example as an extra parameter in the SQL syntax.

cc @adityamaru @stevendanna for triage.

Jira issue: CRDB-20522

Epic CRDB-18749

### PR DESCRIPTION
**Pull request by knz**
Fixes #89967

tldr: multiple changes
- new `WITH tenant_name` clause of `RESTORE TENANT` to set the
  tenant name and ID auto-allocation (auto-allocation is skipped
  if `WITH tenant` is also specified.
- most TENANT SQL syntax can refer to tenants by-name or by-ID.
- `TestServer` can start tenant servers either by-name or by-id.
- builtin `rename_tenant` was removed.
- new syntax form `DROP TENANT ... IMMEDIATE`.
- builtin `destroy_tenant` was deprected in favor of `DROP TENANT [IMMEDIATE]`.

----

Prior to this patch, certain tenant controlling statements were
allowing the user to refer to tenants by numeric ID; whereas others
were referring to tenants only by name.

This commit fixes this by introducing a consistent syntax:

- by default a name is expected;
- if the user wants to refer to a tenant by ID, they can include
  the ID between square brackets. This is consistent with
  other uses of square brackets to refer to tables/databases.

| Statement                                 | Before       | After        |
|-------------------------------------------|--------------|--------------|
| `DROP TENANT <t>`                         | by-name only | both         |
| `CREATE TENANT <t>`                       | by-name only | by-name only |
| `CREATE TENANT <t> FROM REPLICATION OF`   | by-name only | by-name only |
| `CREATE TENANT FROM REPLICATION OF <t>`   | by-name only | by-name only |
| `SHOW TENANT <t>`                         | by-name only | both         |
| `SHOW TENANT <t> WITH REPLICATION STATUS` | by-name only | both         |
| `ALTER TENANT <t> PAUSE REPLICATION`      | by-name only | both         |
| `ALTER TENANT <t> RESUME REPLICATION`     | by-name only | both         |
| `ALTER TENANT <t> COMPLETE REPLICATION`   | by-name only | both         |
| `ALTER TENANT <t> SET REPLICATION`        | by-name only | both         |
| `ALTER TENANT <t> SET CLUSTER SETTING`    | by-ID only   | both         |
| `ALTER TENANT <t> RESET CLUSTER SETTING`  | by-ID only   | both         |
| `ALTER TENANT <t> RENAME TO`              | N/A          | both         |

Additionally, the RETENTION option to CREATE TENANT / ALTER TENANT now
accepts any INTERVAL value, not just strings.

Additionally, it now supports removing the name of a tenant via the
functions `crdb_internal.rename_tenant()` and the new `ALTER TENANT
RENAME TO` statement.

Additionally, it modifies RESTORE TENANT to rename the target tenant,
to avoid name conflits.

Additionally, it makes TestServer able to start a tenant server
designated either by name or by ID.

Release note: None
Epic: CRDB-18749

### CODE REVIEW

**Comment by knz** [comment]
@adityamaru @lidorcarmel you probably want to learn about this.

**Comment by stevendanna** [comment]
:thinking_face: I'm wondering if we really want CREATE TENANT to be able to use an ID.  While it would allow us to support CCs current usage without a crdb_internal builtin, is that worth exposing this ability more generally? We don't allow users to choose the IDs of other objects.

**Comment by knz** [comment]
> I'm wondering if we really want CREATE TENANT to be able to use an ID. While it would allow us to support CCs current usage without a crdb_internal builtin, is that worth exposing this ability more generally? 

Ok - let's not do it.

**Comment by knz** [comment]
ok i removed it

File: pkg/sql/logictest/testdata/logic_test/tenant [hunk 1]
```diff
@@ -144,15 +144,15 @@ statement ok
...
 statement ok
 set default_transaction_read_only = off;
 
 user testuser
 
 statement error only users with the admin role are allowed to destroy tenant
-DROP TENANT "not-allowed"
+DROP TENANT three
 
 user root
```
Comment (stevendanna) [inline pkg/sql/logictest/testdata/logic_test/tenant:157]: Should we add some cases for the ALTER TENANT ... RENAME TO command?

File: pkg/sql/logictest/testdata/logic_test/cluster_settings [hunk 2]
```diff
@@ -213,31 +213,31 @@ subtest tenant-cluster-settings
...
+ALTER TENANT [10] SET CLUSTER SETTING server.mem_profile.total_dump_size_limit='10M'
 
 skipif config 3node-tenant-default-configs
 statement ok
 ALTER TENANT ALL SET CLUSTER SETTING server.mem_profile.total_dump_size_limit='10M'
 
 skipif config 3node-tenant-default-configs
 statement ok
-ALTER TENANT 10 RESET CLUSTER SETTING server.mem_profile.total_dump_size_limit
+ALTER TENANT [10] RESET CLUSTER SETTING server.mem_profile.total_dump_size_limit
```
Comment (stevendanna) [inline pkg/sql/logictest/testdata/logic_test/cluster_settings:240]: Do we want to add a test case here that shows you can set and reset by name as well?

File: pkg/sql/sem/tree/stmt.go [hunk 3]
```diff
@@ -460,6 +460,15 @@ func (*AlterTenantReplication) StatementTag() string { return "ALTER TENANT REPL
 
 func (*AlterTenantReplication) cclOnlyStatement() {}
 
+// StatementReturnType implements the Statement interface.
+func (*AlterTenantRename) StatementReturnType() StatementReturnType { return Rows }
+
+// StatementType implements the Statement interface.
+func (*AlterTenantRename) StatementType() StatementType { return TypeDML }
```
Comment (stevendanna) [inline pkg/sql/sem/tree/stmt.go:467]: Since this is isn't a TypeDDL, I think that we may also want to update CanWriteData to ensure that this can't be run in read-only transactions.

File: pkg/sql/sem/eval/deps.go [hunk 4]
```diff
@@ -536,17 +536,14 @@ type ChangefeedState interface {
...
-	// CreateTenant attempts to install a new tenant in the system and returns the
-	// ID that is assigned to the tenant. It returns an error if another tenant
-	// with `tenantName` already exists. The new tenant is created at the current
-	// active version of the cluster performing the create.
-	CreateTenant(ctx context.Context, tenantName roachpb.TenantName) (roachpb.TenantID, error)
+	// CreateTenant attempts to install a new tenant in the system.
+	// If the provided tenantID is non-zero, it will be used for the new tenant;
+	// an error will be returned if a tenant with the same ID already exists.
+	// Otherwise, a new tenant ID is allocated.
+	// If the provided enantName is non-empty, it will be used for the new tenant.
```
Comment (stevendanna) [inline pkg/sql/sem/eval/deps.go:543]:
```suggestion
	// If the provided tenantName is non-empty, it will be used for the new tenant.
```

File: pkg/sql/sem/tree/walk.go [hunk 5]
```diff
@@ -949,18 +964,149 @@ func (n *AlterTenantSetClusterSetting) walkStmt(v Visitor) Statement {
...
+		if ret == n {
+			ret = n.copyNode()
+		}
+		ret.TenantSpec = ts
+	}
+	return ret
+}
+
+// copyNode makes a copy of this Statement without recursing in any child Statements.
+func (n *AlterTenantReplication) copyNode() *AlterTenantReplication {
```
Comment (stevendanna) [inline pkg/sql/sem/tree/walk.go:978]: Nice catch on this.

File: pkg/sql/sem/eval/deps.go [hunk 6]
```diff
@@ -536,17 +536,14 @@ type ChangefeedState interface {
...
+	// The new tenant is created at the current active version of the
+	// cluster performing the create.
+	CreateTenant(ctx context.Context, tenantID uint64, tenantName roachpb.TenantName) (roachpb.TenantID, error)
```
Comment (stevendanna) [inline pkg/sql/sem/eval/deps.go:546]: Not sure how I feel about using 0 as a sentinel value here.  But, I don't feel strongly about it.

**Review by stevendanna — COMMENTED** [review]
Overall, I like the `[ID]` syntax as that matches what we allow for table IDs in a few places.

I think the relatively small number of places where we needed to change existing tests is a bit of evidence that this won't be too disruptive of a syntax change.

**Review by ecwall — COMMENTED** [review]
Is it possible to make `tenantSpec` into an interface and have 2 impls `tenantSpecSingle`, `tenantSpecAll`. The interface would contain functions that are common to both and the impls would contain the functions that are only relevant to one.

It is hard to know what a function does without reading the source code because in some cases it applies to both impls and in other cases it would be a runtime error to call it with one of them.

> Reviewed 33 of 39 files at r1, 7 of 10 files at r2, 6 of 6 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-:-NLR93ikEhr573aI15Ee:br7hhjc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @renatolabs, and @smg260)

**Review by knz — COMMENTED** [review]
Rebased and updated. RFAL.

> Is it possible to make tenantSpec into an interface and have 2 impls

This was a pretty good idea. Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-:-NLfexUF69yUL-bUbl6g:b7666qa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @renatolabs, @smg260, and @stevendanna)

___
*[`pkg/sql/sem/eval/deps.go` line 543 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMht3XQmjMLn38JoGh-r4-543:-NLfXEYqC-h0SGYqYlpW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/eval/deps.go#L543)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
	// If the provided tenantName is non-empty, it will be used for the new tenant.
```
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/deps.go` line 546 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMht3XQmjMLn38JoGh-r4-546:-NLfYUaJ4JyGtjkL-ARe:b-6zgzze) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/eval/deps.go#L546)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Not sure how I feel about using 0 as a sentinel value here.  But, I don't feel strongly about it.
</blockquote></details>

Do you have a counter-proposal?
___
*[`pkg/sql/sem/tree/stmt.go` line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMLGSRKfk-7ZPNaEzX-r4-467:-NLfZ9BL0XJdI5Gne99G:b-xvll48) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/tree/stmt.go#L467)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Since this is isn't a TypeDDL, I think that we may also want to update CanWriteData to ensure that this can't be run in read-only transactions.
</blockquote></details>

I think they should be TypeDCL, like the one that changes cluster settings.
___
*[`pkg/sql/logictest/testdata/logic_test/cluster_settings` line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLQrUr3Gdtv3Otm6KTB-r4-240:-NLfbIkB3U6ebmR_GoSV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/logictest/testdata/logic_test/cluster_settings#L240)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we want to add a test case here that shows you can set and reset by name as well?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/tenant` line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLQkEUZpZf28o0bqv6g-r4-157:-NLfckiq1kNzzk1hOROZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/logictest/testdata/logic_test/tenant#L157)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we add some cases for the ALTER TENANT ... RENAME TO command?
</blockquote></details>

Done.

**Review by stevendanna — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94932)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, @renatolabs, and @smg260)

___
*[`pkg/sql/tenant_spec.go` line 139 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLg9tU39jJ9LAoqrlkv:-NLg9tU39jJ9LAoqrlkw:bpak3xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccddf0edd58f3e93f921de91da9cf7b997745930/pkg/sql/tenant_spec.go#L139)):*
> ```Go
> 	if ts.All {
> 		return nil, errors.AssertionFailedf("programming error: cannot use ALL in this context")
> 	}
> ```

If fine with this early check, but I think it could be removed since you are going to get this error from getTenantInfo() below.
___
*[`pkg/sql/sem/eval/deps.go` line 546 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMht3XQmjMLn38JoGh-r4-546:-NLg52ZH6eJ0hgBXHrt7:b-mj3xss) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/eval/deps.go#L546)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Do you have a counter-proposal?
</blockquote></details>

I'm OK with it as-is. After thinking about it, perhaps I prefer it.

Here is an explanation of the two thing I don't love about it in case you were interested:

1) The "create-by-id" use case gets first-billing in this API despite us eventually wanting that to be the rare case.

2) The caller needs to traffic in a tenantID that isn't actually a valid tenant ID. But, lots of APIs are this way, like the bind() socket api. And as this PR shows it does make it so a whole bunch of code doesn't need to know the difference between the by-name or by-id case, which is a very nice property.

One alternative would be something where the tenant ID is an optional argument (using the "functional options" pattern that we use in a few places):

```golang
CreateTenant(ctx context.Context, tenantName roachpb.TenantName, opts ...CreateTenantOpts)  (roachpb.TenantID, error)
```

  So a caller that wants to create with a specific ID would do something like:

```golang
CreateTenant(ctx, name, WithTenantID(id))
```

Which to my eye looks nice. You could then have a method on your tenantSpec interface that returned the relevant create options for that tenantSpec. Another path would be a custom type. But that is a lot of "stuff" for what is currently pretty straightforward.
___
*[`pkg/sql/sem/tree/stmt.go` line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMLGSRKfk-7ZPNaEzX-r4-467:-NLgALk70LBmWxQqJLgZ:b-6kb7b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/tree/stmt.go#L467)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I think they should be TypeDCL, like the one that changes cluster settings.
</blockquote></details>

Ah cool, that works for me.

Not for this PR, but I wonder if we should review the other new statements that we've added and see if if they are of the correct type. 
___
*[`pkg/sql/logictest/testdata/logic_test/tenant` line 95 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLgBkiL65M5S-CWHcLn:-NLgBkiM1p5HOzOGUnSu:bpi7vii) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccddf0edd58f3e93f921de91da9cf7b997745930/pkg/sql/logictest/testdata/logic_test/tenant#L95)):*
> ```
> 
> statement ok
> ALTER TENANT blix RENAME TO three
> ```

I could see adding another SHOW TENANTS statement here as a sort of validation that the name was changed.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94932)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @mgartner, @renatolabs, @smg260, and @stevendanna)

___
*[`pkg/sql/tenant_spec.go` line 139 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLg9tU39jJ9LAoqrlkv:-NLgDvM06Hu0BxjBfM5t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccddf0edd58f3e93f921de91da9cf7b997745930/pkg/sql/tenant_spec.go#L139)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

If fine with this early check, but I think it could be removed since you are going to get this error from getTenantInfo() below.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/eval/deps.go` line 546 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMht3XQmjMLn38JoGh-r4-546:-NLqAIbS2zOAPGEzZPsB:btgfeiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/eval/deps.go#L546)):*
> using the "functional options" pattern

So in Go you can't use this pattern when the function is part of an interface. (Or, rather, if you want to use it then you need an OptionsFactory interface able to manufacture options for each type implementing the base interface. This is overkill.)

Instead, I went the route of `generate_test_objects` which also has a compact configuration mechanism. I hope you like it.
___
*[`pkg/sql/sem/tree/stmt.go` line 467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLMLGSRKfk-7ZPNaEzX-r4-467:-NLgDxGL6zkSjht3cAvC:b-ab0w7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/662ecc6f7d8a20271335cc1f0e4c94a30e74034c/pkg/sql/sem/tree/stmt.go#L467)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Ah cool, that works for me.

Not for this PR, but I wonder if we should review the other new statements that we've added and see if if they are of the correct type. 
</blockquote></details>

I just did :)
___
*[`pkg/sql/logictest/testdata/logic_test/tenant` line 95 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/94932#-NLgBkiL65M5S-CWHcLn:-NLgECxb0X9JDy36HOBQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccddf0edd58f3e93f921de91da9cf7b997745930/pkg/sql/logictest/testdata/logic_test/tenant#L95)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I could see adding another SHOW TENANTS statement here as a sort of validation that the name was changed.
</blockquote></details>

Done.

**Comment by knz** [comment]
This PR has been iterating for long enough as it is. Merging now; we can add fixups in followups.

bors r=stevendanna
